### PR TITLE
spirv-val: Add Vulkan Image Format check

### PR DESCRIPTION
**Pull request by spencer-lunarg**
Second attempt at https://github.com/KhronosGroup/SPIRV-Tools/pull/4748

checks for `VUID-StandaloneSpirv-Image-04965`

### CODE REVIEW

File: test/val/val_image_test.cpp [hunk 1]
```diff
@@ -8046,6 +8038,302 @@ TEST_F(ValidateImage, ImageMSArray_ArrayedTypeDoesNotRequireCapability) {
...
+  %u32vec2 = OpTypeVector %u32 2
+  %i32vec4 = OpTypeVector %i32 4
+%u32vec2_01 = OpConstantComposite %u32vec2 %u32_0 %u32_1
+ %img_type = OpTypeImage %i32 2D 0 0 0 2 R32ui
+%ptr_image = OpTypePointer UniformConstant %img_type
+%uniform_image = OpVariable %ptr_image UniformConstant
+     %main = OpFunction %void None %func
+    %label = OpLabel
+      %img = OpLoad %img_type %uniform_image
+     %res1 = OpImageRead %i32vec4 %img %u32vec2_01 ZeroExtend
```
Comment (alan-baker) [inline test/val/val_image_test.cpp:8235]: This seems missing from current validation checks, but the description of ZeroExtend says:
>The texel value is converted to the target value via zero extension. Only valid if the result type is a scalar or vector of [integer type](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#Integer) with signedness of 0.

So this use is invalid.

File: test/val/val_image_test.cpp [hunk 2]
```diff
@@ -8046,6 +8038,302 @@ TEST_F(ValidateImage, ImageMSArray_ArrayedTypeDoesNotRequireCapability) {
...
+  EXPECT_THAT(getDiagnosticString(),
+              AnyVUID("VUID-StandaloneSpirv-Image-04965"));
+  EXPECT_THAT(getDiagnosticString(),
+              HasSubstr("Can't use SignExtend or ZeroExtend on float images"));
+}
+
+TEST_F(ValidateImage, BothSignExtendZeroExtendTogether) {
+  const std::string body = R"(
+%img = OpLoad %type_image_s32_2d_0002 %uniform_image_s32_2d_0002
+%res1 = OpImageRead %s32vec4 %img %u32vec2_01 SignExtend ZeroExtend
```
Comment (alan-baker) [inline test/val/val_image_test.cpp:8326]: I think this is written incorrectly. It should be:
```suggestion
%res1 = OpImageRead %s32vec4 %img %u32vec2_01 SignExtend|ZeroExtend
```
Right now it is two operands and the parser is complaining. Though I admit it seems strange that both would be allowed, I can't find where it is forbidden. Maybe that should be a spec issue.

**Review by alan-baker — CHANGES_REQUESTED** [review]

File: source/val/validate_image.cpp [hunk 3]
```diff
@@ -178,6 +178,97 @@ bool IsValidGatherLodBiasAMD(const ValidationState_t& _, spv::Op opcode) {
   return false;
 }
 
+bool IsImageFormatInt32Width(spv::ImageFormat format) {
```
[the remark this answers is not in the document]
Comment (spencer-lunarg) [reply source/val/validate_image.cpp:181]: I guess I'm slightly confused why you want `Sampled Type` here when it is checking the `Image Format` operand (`ImageTypeInfo::format`)

File: test/val/val_image_test.cpp (hunk 2 above)
Comment (spencer-lunarg) [reply test/val/val_image_test.cpp:8326]: going to remove this test and come back later with it once we get spec text (also it is slightly out of scope of this PR)

File: source/val/validate_image.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (spencer-lunarg) [reply source/val/validate_image.cpp]: fair enough, will do!

[the remark this answers is not in the document]
Comment (spencer-lunarg) [reply source/val/validate_image.cpp]: sorry, looking again I realized these name are fully misleading and just wrongly named, think new commit clears it up

File: source/val/validate_image.cpp [hunk 4]
```diff
@@ -634,6 +759,16 @@ spv_result_t ValidateImageOperands(ValidationState_t& _,
...
+    if (!_.IsIntScalarOrVectorType(result_type)) {
+      return _.diag(SPV_ERROR_INVALID_DATA, inst)
+             << _.VkErrorID(4965)
+             << "Using SignExtend, but result type is not a scalar or vector "
+                "integer type.";
+    } else if (!_.IsUnsignedIntScalarOrVectorType(result_type)) {
+      return _.diag(SPV_ERROR_INVALID_DATA, inst)
+             << _.VkErrorID(4965)
+             << "Using SignExtend, but result type is a signed integer type.";
+    }
```
Comment (alan-baker) [inline source/val/validate_image.cpp:771]: We know that ZeroExtend is set, so let's just check for an unsigned int. Also the wrong operand value is listed in the error.

```suggestion
   if (!_.IsUnsignedIntScalarOrVectorType(result_type)) {
      return _.diag(SPV_ERROR_INVALID_DATA, inst)
             << _.VkErrorID(4965)
             << "Using ZeroExtend, but result type is a signed integer type.";
    }
```

**Review by alan-baker — CHANGES_REQUESTED** [review]

**Comment by alan-baker** [comment]
There are several glslang unit test failures. 

The sparse failures look like a spec problem. Sparse samples return a struct with residency and value. The zero/sign extend should apply to the second element.

The image write case is a similar problem, spec says result type, but there is no result type.

I'm unsure about the HLSL test. It looks incorrect to me, but I'm not sure what the strategy for specifying formats is when compiling HLSL.

**Comment by alan-baker** [comment]
See Khronos internal issue 766.

**Comment by spencer-lunarg** [comment]
with https://gitlab.khronos.org/spirv/SPIR-V/-/merge_requests/302 merged, will apply and rebase

**Comment by spencer-lunarg** [comment]
> I believe we are still waiting for a rebase so we Alan can review again. Is that correct?

yes, need to apply spec change. Honestly happy to close and create a new fresh PR if desired, otherwise will make changes here

**Comment by s-perron** [comment]
Sure open a new one. Thanks.

**Pull request closed without merging**